### PR TITLE
Scale how the data is downloaded in get_data.py

### PR DESCRIPTION
**Pull request by umohnani8**
If there are more than 1 pod, the init-container of one pod should download
the data while the init-containers of the other pods wait. Once the data is downloaded
the processing container of all the pods can access the data.

Signed-off-by: umohnani8 <umohnani@redhat.com>

### CODE REVIEW

**Comment by umohnani8** [comment]
Addresses issue https://github.com/redhat-university-partnerships/radiology/issues/45#event-1478636684
@danmcp @ravisantoshgudimetla PTAL
I can remove the print statements if they are not needed. I just think that they would be helpful to be in the logs for the user if something goes wrong.

File: openshift/pman-swift-publisher/Dockerfile [hunk 1]
```diff
@@ -17,7 +17,8 @@ RUN apt-get update \
     && apt-get install -y curl \
     && pip3 install keystoneauth1 \ 
     && pip3 install python-swiftclient \
-    && pip3 install kubernetes
+    && pip3 install kubernetes \
+    && pip3 install lockfile
```
Comment (ravisantoshgudimetla) [inline openshift/pman-swift-publisher/Dockerfile]: Seems this package has been deprecated. (https://pypi.python.org/pypi/lockfile).

File: openshift/pman-swift-publisher/get_data.py [hunk 2]
```diff
@@ -64,9 +66,27 @@ def getData(self, **kwargs):
...
             os.makedirs('/share/outgoing')
 
-        
 
 if __name__ == "__main__":
 
     obj = SwiftStore()
-    obj.getData(path= os.environ.get('SWIFT_KEY'))
\ No newline at end of file
+    try:
```
Comment (ravisantoshgudimetla) [inline openshift/pman-swift-publisher/get_data.py]: Separating this as function might help @husky-parul. Not sure if she is following the same method though.

File: openshift/pman-swift-publisher/get_data.py [hunk 3]
```diff
@@ -64,9 +66,27 @@ def getData(self, **kwargs):
...
+        lock = LockFile('/share/lockfile')
+        # if the lock file doesn't exist, check if /share is empty also (this means that the data has not been downloaded yet)
+        # checking /share because /share/incoming does not exist yet and /share is where the shared directory is mounted onto
+        # /share should always be empty when the init-container starts
+        if not lock.is_locked() and os.listdir('/share') == []:
+            print("I am master. Locking incoming directory and downloading data")
+            lock.acquire()
```
Comment (ravisantoshgudimetla) [inline openshift/pman-swift-publisher/get_data.py]: nit: What will happen if lock is acquired by code in pod1 and before downloading if it fails then and lock is not released? Will the download happen again? If not, that is fine too. Was just curious but please add a TODO.

File: openshift/pman-swift-publisher/Dockerfile (hunk 1 above)
Comment (umohnani8) [reply openshift/pman-swift-publisher/Dockerfile]: Oh, I didn't see that. Which package would be better to use, fasteners or oslo.concurrency?

Comment (ravisantoshgudimetla) [reply openshift/pman-swift-publisher/Dockerfile]: I don't know about fasteners but oslo is well maintained in OpenStack community.

Comment (danmcp) [reply openshift/pman-swift-publisher/Dockerfile]: Oslo is probably too big for us to pull in.

File: openshift/pman-swift-publisher/get_data.py (hunk 3 above)
Comment (danmcp) [reply openshift/pman-swift-publisher/get_data.py]: This is a great catch although I wouldn't call it a nit.  To solve you might want to do something like:

- catch the exception from a failure to download
- write a second file indicating a failure to download
- reraise the exception so the pod exits with failure
- when the next pod starts up to replace, there needs to be code to see the extra file indicating a failure to download.  At which point it removes that file and it is now the owner of the lock and is responsible for the download.

Comment (ravisantoshgudimetla) [reply openshift/pman-swift-publisher/get_data.py]: Another point 
  - what will happen if the code crashes in the middle. Pod1 acquired lock and crashed before releasing it. 

We need to have timeout to break the infinite loop.

File: openshift/pman-swift-publisher/get_data.py [hunk 4]
```diff
@@ -64,9 +66,27 @@ def getData(self, **kwargs):
...
 
     obj = SwiftStore()
-    obj.getData(path= os.environ.get('SWIFT_KEY'))
\ No newline at end of file
+    try:
+        lock = LockFile('/share/lockfile')
+        # if the lock file doesn't exist, check if /share is empty also (this means that the data has not been downloaded yet)
+        # checking /share because /share/incoming does not exist yet and /share is where the shared directory is mounted onto
+        # /share should always be empty when the init-container starts
+        if not lock.is_locked() and os.listdir('/share') == []:
```
[the remark this answers is not in the document]
Comment (ravisantoshgudimetla) [reply openshift/pman-swift-publisher/get_data.py]: As long as we use a context manager, we should solve the problem. Code would look like:

```Python
try: 
    with LockFile("/share/lockfile"): # We have acquired lock now. This would throw an exception if it is already locked. This ensures atomicity of operation.
         try:  
              "File download if /share is empty"
         except: # Catch File Download exception.
              release lock so that next pod could acquire lock and start downloading. 
    except:
        "Lock acquiring failed meaning some other process is still using the lock file (or) filedownload failed. Wait for sometime and eventually exit after n hours/min"
```
@danmcp, @umohnani8 - I might have missed some other edge case but WDYT?

[the remark this answers is not in the document]
Comment (danmcp) [reply openshift/pman-swift-publisher/get_data.py]: Was the listdir call not necessary before?

Nit: filedownload should be two words in the except

[the remark this answers is not in the document]
Comment (ravisantoshgudimetla) [reply openshift/pman-swift-publisher/get_data.py]: > Was the listdir call not necessary before?

I think it is needed for other pods to not download even if they acquire lock.

[the remark this answers is not in the document]
Comment (danmcp) [reply openshift/pman-swift-publisher/get_data.py]: Is needed or was needed?

[the remark this answers is not in the document]
Comment (ravisantoshgudimetla) [reply openshift/pman-swift-publisher/get_data.py]: Do you mean with the way I am proposing is it needed? Yes it is still needed and it was needed back then as well.

[the remark this answers is not in the document]
Comment (danmcp) [reply openshift/pman-swift-publisher/get_data.py]: How do 2 pods acquire the lock with this logic to make the listdir call necessary?

**Comment by umohnani8** [comment]
@ravisantoshgudimetla I updated the code, PTAL.

File: openshift/pman-openshift-template.json [hunk 5]
```diff
@@ -76,7 +76,7 @@
                   "cpu": "2000m"
                 },
                 "requests": {
-                  "memory": "128Mi",
+                  "memory": "150Mi",
```
Comment (ravisantoshgudimetla) [inline openshift/pman-openshift-template.json]: Why do we need to update this? If needed, please provide a comment.

File: openshift/pman-swift-publisher/Dockerfile [hunk 6]
```diff
@@ -17,7 +17,8 @@ RUN apt-get update \
     && apt-get install -y curl \
     && pip3 install keystoneauth1 \ 
     && pip3 install python-swiftclient \
-    && pip3 install kubernetes
+    && pip3 install kubernetes \
+    && pip3 install fasteners
```
Comment (ravisantoshgudimetla) [inline openshift/pman-swift-publisher/Dockerfile]: I think we need to update this as well https://github.com/FNNDSC/pman/blob/master/setup.py#L21 to work for non-containerized versions. You can do it in a follow-up PR. Please add a TODO.

File: openshift/pman-swift-publisher/get_data.py [hunk 7]
```diff
@@ -64,9 +66,34 @@ def getData(self, **kwargs):
...
+    # the init-storage container in all the pods should acquire the lock
+    with fasteners.InterProcessLock("/share/lockfile"):
+        # download the data if "share/incoming" is empty or does not exist
+        if not has_data():
+            try:
+                print("Downloading data from swift...")
+                obj.getData(path=os.environ.get('SWIFT_KEY'))
+            except Exception as err:
+                print("Failed to download the data:", err)
+                pass
```
Comment (ravisantoshgudimetla) [inline openshift/pman-swift-publisher/get_data.py]: Why pass here?

File: openshift/pman-swift-publisher/get_data.py [hunk 8]
```diff
@@ -64,9 +66,34 @@ def getData(self, **kwargs):
...
     obj = SwiftStore()
-    obj.getData(path= os.environ.get('SWIFT_KEY'))
\ No newline at end of file
+    start_time = time.time()
+    # the init-storage container in all the pods should acquire the lock
+    with fasteners.InterProcessLock("/share/lockfile"):
+        # download the data if "share/incoming" is empty or does not exist
+        if not has_data():
+            try:
+                print("Downloading data from swift...")
```
Comment (ravisantoshgudimetla) [inline openshift/pman-swift-publisher/get_data.py]: nit: Can you add something like, "Acquiring lock to download"

File: openshift/pman-swift-publisher/get_data.py [hunk 9]
```diff
@@ -64,9 +66,34 @@ def getData(self, **kwargs):
...
+                obj.getData(path=os.environ.get('SWIFT_KEY'))
+            except Exception as err:
+                print("Failed to download the data:", err)
+                pass
+    # wait for download to finish before moving on to the next container
+    while not has_data():
+        # if the pods have been trying to download the data for more than 20 minutes
+        # that means something is wrong, so exit with error
+        if time.time() - start_time > 1200:
```
Comment (ravisantoshgudimetla) [inline openshift/pman-swift-publisher/get_data.py]: nit: Can you add a print here?

File: pman/openshiftmgr.py [hunk 10]
```diff
@@ -192,7 +192,8 @@ def schedule(self, image, command, name, number_of_workers,
             d_job['spec']['template']['spec']['volumes'] = [
                 {
                     "name": "shared-volume",
-                    "emptyDir": {
+                    "persistentVolumeClaim": {
+                        "claimName": name + "-storage-claim"
```
Comment (ravisantoshgudimetla) [inline pman/openshiftmgr.py]: So if IIUC, the assumption is that claimName is "name" of the job appended with "-storage-claim"? If yes, can you please add a comment saying so and create a PR for it.

File: openshift/pman-openshift-template.json (hunk 5 above)
Comment (umohnani8) [reply openshift/pman-openshift-template.json]: Because the minimum memory allowed by MOC is 150Mi, that is why I updated the templates to reflect that.

File: openshift/pman-swift-publisher/Dockerfile (hunk 6 above)
Comment (umohnani8) [reply openshift/pman-swift-publisher/Dockerfile]: Added it there.

File: openshift/pman-swift-publisher/get_data.py (hunk 7 above)
Comment (umohnani8) [reply openshift/pman-swift-publisher/get_data.py]: Not needed, removed.

File: openshift/pman-swift-publisher/get_data.py (hunk 8 above)
Comment (umohnani8) [reply openshift/pman-swift-publisher/get_data.py]: Added.

File: openshift/pman-swift-publisher/get_data.py (hunk 9 above)
Comment (umohnani8) [reply openshift/pman-swift-publisher/get_data.py]: Removed timeout check from here. Will add in openshiftmgr.py

File: pman/openshiftmgr.py (hunk 10 above)
Comment (umohnani8) [reply pman/openshiftmgr.py]: yup, adding in openshiftmgr.py

**Comment by ravisantoshgudimetla** [comment]
@umohnani8 Seems something broke in tests. Can you please check that otherwise this LGTM.

**Comment by ravisantoshgudimetla** [comment]
@danmcp PTAL

File: openshift/pman-swift-publisher/get_data.py [hunk 11]
```diff
@@ -64,9 +66,29 @@ def getData(self, **kwargs):
...
             os.makedirs('/share/outgoing')
 
-        
 
-if __name__ == "__main__":
+# has_data returns true if path exists and is not empty
+def has_data(path='/share/incoming'):
+    if not os.path.exists(path):
+        return False
+    if os.listdir(path) == []:
```
Comment (danmcp) [inline openshift/pman-swift-publisher/get_data.py]: use len() == 0

File: openshift/pman-swift-publisher/get_data.py [hunk 12]
```diff
@@ -64,9 +66,29 @@ def getData(self, **kwargs):
...
+            try:
+                print("Lock acquired. Downloading data from swift...")
+                obj.getData(path=os.environ.get('SWIFT_KEY'))
+            except Exception as err:
+                print("Failed to download the data:", err)
+                exit(1)
+    # wait for download to finish before moving on to the next container
+    while not has_data():
+        time.sleep(2)
+    print("Data Downloaded!")
```
Comment (danmcp) [inline openshift/pman-swift-publisher/get_data.py]: use lower case d on downloaded

File: openshift/pman-swift-publisher/get_data.py [hunk 13]
```diff
@@ -64,9 +66,29 @@ def getData(self, **kwargs):
...
+if __name__ == "__main__":
     obj = SwiftStore()
-    obj.getData(path= os.environ.get('SWIFT_KEY'))
\ No newline at end of file
+    # the init-storage container in all the pods should acquire the lock
+    with fasteners.InterProcessLock("/share/lockfile"):
+        # download the data if "share/incoming" is empty or does not exist
+        if not has_data():
+            try:
+                print("Lock acquired. Downloading data from swift...")
```
Comment (danmcp) [inline openshift/pman-swift-publisher/get_data.py]: capital S in swift.

File: openshift/pman-swift-publisher/get_data.py [hunk 14]
```diff
@@ -64,9 +66,29 @@ def getData(self, **kwargs):
...
+    if os.listdir(path) == []:
+        return False
+    return True
+
 
+if __name__ == "__main__":
     obj = SwiftStore()
-    obj.getData(path= os.environ.get('SWIFT_KEY'))
\ No newline at end of file
+    # the init-storage container in all the pods should acquire the lock
```
Comment (danmcp) [inline openshift/pman-swift-publisher/get_data.py]: Start sentences in comments with capital letters

File: openshift/pman-swift-publisher/get_data.py (hunk 11 above)
Comment (umohnani8) [reply openshift/pman-swift-publisher/get_data.py]: Fixed

File: openshift/pman-swift-publisher/get_data.py (hunk 12 above)
Comment (umohnani8) [reply openshift/pman-swift-publisher/get_data.py]: Fixed

File: openshift/pman-swift-publisher/get_data.py (hunk 13 above)
Comment (umohnani8) [reply openshift/pman-swift-publisher/get_data.py]: Fixed

File: openshift/pman-swift-publisher/get_data.py (hunk 14 above)
Comment (umohnani8) [reply openshift/pman-swift-publisher/get_data.py]: Fixed

File: pman/openshiftmgr.py [hunk 15]
```diff
@@ -192,7 +192,8 @@ def schedule(self, image, command, name, number_of_workers,
             d_job['spec']['template']['spec']['volumes'] = [
                 {
                     "name": "shared-volume",
-                    "emptyDir": {
+                    "persistentVolumeClaim": {
```
[the remark this answers is not in the document]
Comment (umohnani8) [reply pman/openshiftmgr.py]: Added the functionality to create the PVC.

**Comment by umohnani8** [comment]
Should be ready now, @ravisantoshgudimetla @danmcp PTAL.

**Comment by umohnani8** [comment]
Storage is hard coded to 5Gi for now, I think that should be something the user should be able to control, but can add that functionality in later.

File: openshift/pman-openshift-template.json (hunk 5 above)
Comment (ravisantoshgudimetla) [reply openshift/pman-openshift-template.json]: Add a comment somewhere saying that this is specific to MOC.

File: openshift/pman-swift-publisher/Dockerfile (hunk 6 above)
Comment (ravisantoshgudimetla) [reply openshift/pman-swift-publisher/Dockerfile]: Thanks.

File: openshift/pman-swift-publisher/get_data.py [hunk 16]
```diff
@@ -64,9 +66,29 @@ def getData(self, **kwargs):
...
+    return True
+
 
+if __name__ == "__main__":
     obj = SwiftStore()
-    obj.getData(path= os.environ.get('SWIFT_KEY'))
\ No newline at end of file
+    # The init-storage container in all the pods should acquire the lock
+    with fasteners.InterProcessLock("/share/lockfile"):
+        # Download the data if "share/incoming" is empty or does not exist
```
Comment (ravisantoshgudimetla) [inline openshift/pman-swift-publisher/get_data.py]: `"/share/incoming"` is empty

File: pman/openshiftmgr.py [hunk 17]
```diff
@@ -339,3 +340,29 @@ def get_pod_names_in_job(self, job_id):
...
+                "accessModes": [
+                    "ReadWriteMany"
+                ],
+                "resources": {
+                    "requests": {
+                        "storage": "5Gi"
+                    }
+                }
+            }
+        }
```
Comment (ravisantoshgudimetla) [inline pman/openshiftmgr.py]: Probably try-catch may be better, in case pvc creation fails

File: openshift/pman-openshift-template.json (hunk 5 above)
Comment (umohnani8) [reply openshift/pman-openshift-template.json]: Added comment in the README

File: openshift/pman-swift-publisher/get_data.py (hunk 16 above)
Comment (umohnani8) [reply openshift/pman-swift-publisher/get_data.py]: Fixed

File: pman/openshiftmgr.py [hunk 18]
```diff
@@ -339,3 +340,29 @@ def get_pod_names_in_job(self, job_id):
...
+        pvc = self.kube_client.create_namespaced_persistent_volume_claim(self.project, body=d_pvc)
+        return pvc
```
[the remark this answers is not in the document]
Comment (umohnani8) [reply pman/openshiftmgr.py]: I guess not. I was following the way the schedule and create_pod functions were written.

**Comment by umohnani8** [comment]
@danmcp @ravisantoshgudimetla should the pvc deletion happen here https://github.com/FNNDSC/pman/blob/master/pman/pman.py#L1360?
If so, how do I go about testing that? How do I call the def_Shutdown() function?

File: openshift/pman-openshift-template.json (hunk 5 above)
Comment (danmcp) [reply openshift/pman-openshift-template.json]: I would rather not add such a comment.  The value is arbitrary anyway.  I can live with this arbitrary value but I don't want MOC specific statements in the code unless absolutely necessary.

File: pman/openshiftmgr.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (umohnani8) [reply pman/openshiftmgr.py]: Fixed

File: pman/openshiftmgr.py (hunk 17 above)
Comment (umohnani8) [reply pman/openshiftmgr.py]: Fixed

File: openshift/README.rst [hunk 19]
```diff
@@ -26,7 +26,7 @@ Assuming oc cluster up has been run.
     rm -f ~/.kube/config
     oc login # As developer
     # Ignore this step, if you are using swift as backend storage.
-    oc new-app openshift/pman-openshift-template-without-swift.json
+    oc new-app openshift/pman-openshift-template-without-swift.json # The memory request is set to 150Mi as that is the minimum allowed
```
[the remark this answers is not in the document]
Comment (umohnani8) [reply openshift/README.rst]: Removed the comment.

File: pman/pman.py [hunk 20]
```diff
@@ -1361,6 +1361,10 @@ def job_shutDown(d_serviceInfo):
             """
             Shut down a service
             """
+            try:
+                self.get_openshift_manager().remove_pvc(jid)
```
Comment (umohnani8) [inline pman/pman.py]: The only thing I am not sure of is whether the `remove_pvc()` function should be called here.
@danmcp @ravisantoshgudimetla can you please double check, thanks!

**Comment by danmcp** [comment]
@umohnani8 This PR has a merge conflict

File: openshift/pman-swift-publisher/get_data.py [hunk 21]
```diff
@@ -64,9 +66,29 @@ def getData(self, **kwargs):
...
+        # Download the data if "/share/incoming" is empty or does not exist
+        if not has_data():
+            try:
+                print("Lock acquired. Downloading data from Swift...")
+                obj.getData(path=os.environ.get('SWIFT_KEY'))
+            except Exception as err:
+                print("Failed to download the data:", err)
+                exit(1)
+    # Wait for download to finish before moving on to the next container
+    while not has_data():
```
Comment (danmcp) [inline openshift/pman-swift-publisher/get_data.py]: @umohnani8 Why is this necessary?  Is InterProcessLock not blocking?  Also, if this is necessary, how is has_data the correct call?  It seems like it only waits for the download to start but not finish.

Comment (ravisantoshgudimetla) [reply openshift/pman-swift-publisher/get_data.py]: This is not needed. We are getting blocked on acquiring lock. But we cannot trust if the file download has happened.

Comment (umohnani8) [reply openshift/pman-swift-publisher/get_data.py]: Fixed this. The while loop wasn't needed as the locking context manager blocks other processes if one process has the lock so removed that. Added code to create a "success" file if the download happens successfully, the other pods' init-container will check for the existence of this file as well to determine whether to download data from swift or not.
@ravisantoshgudimetla @danmcp PTAL

File: openshift/pman-swift-publisher/get_data.py [hunk 22]
```diff
@@ -64,9 +66,35 @@ def getData(self, **kwargs):
...
-    obj.getData(path= os.environ.get('SWIFT_KEY'))
\ No newline at end of file
+    # The init-storage container in all the pods should acquire the lock
+    with fasteners.InterProcessLock("/share/lockfile"):
+        # Download the data if "/share/incoming" is empty or does not exist
+        if not has_data():
+            try:
+                print("Lock acquired. Downloading data from Swift...")
+                obj.getData(path=os.environ.get('SWIFT_KEY'))
+                #create a success file, if download completed successfully
```
Comment (danmcp) [inline openshift/pman-swift-publisher/get_data.py]: Use consistent spacing/capitalization for comments:
# Create

File: openshift/pman-swift-publisher/get_data.py [hunk 23]
```diff
@@ -64,9 +66,35 @@ def getData(self, **kwargs):
...
-        
 
-if __name__ == "__main__":
+# has_data returns true if path exists and is not empty
+def has_data(path='/share/incoming'):
+    if not os.path.exists(path):
+        return False
+    if len(os.listdir(path)) == 0:
+        return False
+    files = os.listdir(path)
```
Comment (danmcp) [inline openshift/pman-swift-publisher/get_data.py]: Can't you just check for the one file existing rather than listing the dir?

File: openshift/pman-swift-publisher/get_data.py [hunk 24]
```diff
@@ -64,9 +66,35 @@ def getData(self, **kwargs):
...
             os.makedirs('/share/outgoing')
 
-        
 
-if __name__ == "__main__":
+# has_data returns true if path exists and is not empty
+def has_data(path='/share/incoming'):
+    if not os.path.exists(path):
+        return False
+    if len(os.listdir(path)) == 0:
```
Comment (danmcp) [inline openshift/pman-swift-publisher/get_data.py]: Do we still need this check and the one above it?  Can't we just check for the success file?

File: openshift/pman-swift-publisher/get_data.py [hunk 25]
```diff
@@ -64,9 +66,35 @@ def getData(self, **kwargs):
...
+    with fasteners.InterProcessLock("/share/lockfile"):
+        # Download the data if "/share/incoming" is empty or does not exist
+        if not has_data():
+            try:
+                print("Lock acquired. Downloading data from Swift...")
+                obj.getData(path=os.environ.get('SWIFT_KEY'))
+                # Create a success file, if download completed successfully
+                try:
+                    with open('/share/incoming/success', 'w'): pass
+                except Exception as err:
```
Comment (danmcp) [inline openshift/pman-swift-publisher/get_data.py]: Why catch the exception?

Comment (danmcp) [reply openshift/pman-swift-publisher/get_data.py]: Or at least why not reraise it?

File: openshift/pman-swift-publisher/get_data.py [hunk 26]
```diff
@@ -64,9 +66,35 @@ def getData(self, **kwargs):
...
+
 
+if __name__ == "__main__":
     obj = SwiftStore()
-    obj.getData(path= os.environ.get('SWIFT_KEY'))
\ No newline at end of file
+    # The init-storage container in all the pods should acquire the lock
+    with fasteners.InterProcessLock("/share/lockfile"):
+        # Download the data if "/share/incoming" is empty or does not exist
+        if not has_data():
```
Comment (danmcp) [inline openshift/pman-swift-publisher/get_data.py]: Should has_data be is_downloaded()?

File: openshift/pman-swift-publisher/get_data.py (hunk 24 above)
Comment (ravisantoshgudimetla) [reply openshift/pman-swift-publisher/get_data.py]: +1 to what Dan suggested.

File: openshift/pman-swift-publisher/get_data.py (hunk 26 above)
Comment (ravisantoshgudimetla) [reply openshift/pman-swift-publisher/get_data.py]: Or the file presence condition.

File: openshift/pman-swift-publisher/get_data.py (hunk 25 above)
Comment (ravisantoshgudimetla) [reply openshift/pman-swift-publisher/get_data.py]: I think, try catch is needed, if the file creation files, in except, you can either 
a.) Delete the downloaded file and exit with failure.
b.) try creating the success file again for n number of times and see if it is successful, if yes, proceed with finally with exit(1) or fail with exit(1).

I think a.) should be sufficient.

**Comment by umohnani8** [comment]
Updated the get_data logic. Tested the various failure and success cases. @ravisantoshgudimetla @danmcp PTAL.

File: pman/pman.py (hunk 20 above)
Comment (danmcp) [reply pman/pman.py]: It should be here but I don't think the error handling is ideal.  I think removing the job and removing the pvc should both be attempted and an error returned if either failed.  Either that or both some be attempted and the error logged if either failed.  I would recommend the former for now and we can consider using a a secondary cleanup mechanism to clean them up if this process misses them.

File: openshift/pman-swift-publisher/get_data.py [hunk 27]
```diff
@@ -64,9 +66,29 @@ def getData(self, **kwargs):
...
\ No newline at end of file
+    # The init-storage container in all the pods should acquire the lock
+    with fasteners.InterProcessLock("/share/lockfile"):
+        # If "download-failed" exists, exit with an error code immediately
+        if os.path.exists("/share/download-failed"):
+            print("Previous pod failed to download the data. Exiting with failure...")
+            exit(1)
+        # If there is some data in "/share/incoming" but "/share/download-succeeded" doesn't exist, it is a failure case
+        # Exit with error code immediately
+        if os.path.exists("/share/incoming") and len(os.listdir('/share/incoming')) > 0 and not os.path.exists('/share/download-succeeded'):
```
Comment (danmcp) [inline openshift/pman-swift-publisher/get_data.py]: We shouldn't be hardcoding /share/incoming and /share/outgoing.  The directories should be passed to the cmd execution and we should be getting them from there.

File: openshift/pman-swift-publisher/get_data.py [hunk 28]
```diff
@@ -64,9 +66,29 @@ def getData(self, **kwargs):
...
 
-        
-
 if __name__ == "__main__":
-
     obj = SwiftStore()
-    obj.getData(path= os.environ.get('SWIFT_KEY'))
\ No newline at end of file
+    # The init-storage container in all the pods should acquire the lock
+    with fasteners.InterProcessLock("/share/lockfile"):
```
Comment (danmcp) [inline openshift/pman-swift-publisher/get_data.py]: Should we use .files for each of these files we are writing?

File: openshift/pman-swift-publisher/get_data.py (hunk 27 above)
Comment (danmcp) [reply openshift/pman-swift-publisher/get_data.py]: We could avoid this by writing a download started file instead of using the incoming dir.  If we can't write that file, it would mean all container would try to write that file and fail, but that's not the end of the world.

Comment (umohnani8) [reply openshift/pman-swift-publisher/get_data.py]: The directories have always been hardcoded. You want me to make the change in this PR or open another one to get rid of the hard-coding?

Wouldn't adding a third file just be adding more complexity. We know we have to check the data in /share/incoming so why not just check that instead of checking if a third file exists, and if it doesn't create it etc etc.

File: openshift/pman-swift-publisher/get_data.py (hunk 28 above)
Comment (umohnani8) [reply openshift/pman-swift-publisher/get_data.py]: Sure, I can add that.

File: openshift/pman-swift-publisher/get_data.py (hunk 27 above)
Comment (danmcp) [reply openshift/pman-swift-publisher/get_data.py]: Yeah I just missed it before.  Previously there were other cases that were hardcoded and this addition would be the only one that still was.

The alternative of using another file is just because getting the directory might be a lot more difficult.  Getting the correct dir is preferable.

Comment (umohnani8) [reply openshift/pman-swift-publisher/get_data.py]: Fixed. Getting the directories from the cmd line.

File: openshift/pman-swift-publisher/get_data.py (hunk 28 above)
Comment (umohnani8) [reply openshift/pman-swift-publisher/get_data.py]: Fixed.

**Comment by umohnani8** [comment]
@danmcp should be ready now.

**Comment by danmcp** [comment]
@umohnani8 Please squash the commits and I am good to merge

File: pman/pman.py [hunk 29]
```diff
@@ -1871,6 +1875,10 @@ def t_run_process_openshift(self, *args, **kwargs):
             self.auid           = d_meta['auid']
             str_cmd             = d_meta['cmd']
 
+            str_arr = str_cmd.split()
+            incoming_dir = str_arr[len(str_arr)-2]
```
Comment (ravisantoshgudimetla) [inline pman/pman.py]: Sorry for nitpicking but order is not guaranteed /share/incoming can come after /share/outgoing. I think inputdir and output are the name of the arguments that we need to look at.

Comment (danmcp) [reply pman/pman.py]: @ravisantoshgudimetla Aren't they positional arguments?

Comment (umohnani8) [reply pman/pman.py]: Yeah, I thought they were positional arguments in the cmd passed in as part of the json to pfurl.

**Comment by umohnani8** [comment]
@danmcp squashed commits.

File: openshift/pman-swift-publisher/get_data.py [hunk 30]
```diff
@@ -56,17 +63,36 @@ def getData(self, **kwargs):
...
+            os.makedirs(outgoing_dir)
 
 if __name__ == "__main__":
-
+    incoming_dir = os.environ.get("INCOMING_DIR")
     obj = SwiftStore()
-    obj.getData(path= os.environ.get('SWIFT_KEY'))
\ No newline at end of file
+    # The init-storage container in all the pods should acquire the lock
+    with fasteners.InterProcessLock("/share/lockfile.file"):
```
Comment (danmcp) [inline openshift/pman-swift-publisher/get_data.py]: Sorry just noticed this, by .file I meant /share/.lockfile or /share/.download-failed

File: pman/pman.py (hunk 29 above)
Comment (ravisantoshgudimetla) [reply pman/pman.py]: @umohnani8 Please ignore my comments. After discussing with @danmcp, we decided it is better to take this in a larger meeting and not block you(Ref: https://github.com/FNNDSC/pman/pull/68/files#diff-4dc555deb0667fb88394550223dee03cL61). 

One recommendation would be guard str_arr better for example, we can check if len(str_arr) > 2 and then proceed further.

File: openshift/pman-swift-publisher/get_data.py (hunk 30 above)
Comment (umohnani8) [reply openshift/pman-swift-publisher/get_data.py]: oh whoops, fixed now.

File: pman/pman.py (hunk 29 above)
Comment (umohnani8) [reply pman/pman.py]: I can change the pfurl json to have `incoming_dir` and `outgoing_dir` keys instead of having the directories at the end of the string given to `cmd`. Not sure if that will cause issues though.

Comment (danmcp) [reply pman/pman.py]: @umohnani8 That's a much bigger change than just changing pfurl.  You would have to change every existing plugin implementation and chrisapp.
